---
title: "R Notebook"
output:
  pdf_document: default
  html_document:
    df_print: paged
editor_options: 
  markdown: 
    wrap: 72
---

## Welcome to Assignment 2, Data Exploration

# Instructions:

[Below we've transcribed the assignment instructions with corresponding
code blocks beneath each.]{.underline}

[üö®Please note,]{.underline}

-   [underlined writing implies it is from us and not the
    instructions]{.underline}

-   [numerous code blocks were directly excerpted, adapted, and reused
    from the provided instructions / guidance. We do not believe this to
    be plagiarism and hope that in declaring it up front we demonstrate
    our honest intent. The sources used are found in the instructions
    (which are transcribed in full on this document), and due to how
    closely the guidance reflected the tasks of the assignment, we felt
    it wholly appropriate to directly use the best practices provided
    directly to us for this purpose.]{.underline}

-   [Task 7: Annotating our tables, has been performed inline, in a more
    traditional notebook manner. We do not reserve the annotations for
    step 7 and instead hope to meet the criteria by elaborating on all
    provided images and tables directly adjacent to each.]{.underline}

-   [‚≠êÔ∏è Stars are used to denote items corresponding directly to the
    rubric]{.underline}

```{r}
# Mount requisite libraries
library(dplyr)
library(tidyverse)
library(readr)
```

### 1. Download the expression data and matching metadata from Refine.Bio that you selected in Assignment 1.

-   You should have a matrix of samples by genes expression data. If
    your matrix has Ensembl IDs (e.g. ENSG00000141510) instead of Hugo
    gene names (e.g. TP53), convert the names following these
    directions:

    -   alexslemonade.github.io/refinebio-examples/03-rnaseq/gene-id-annotation_rnaseq_01_ensembl.html
        bioconductor.org/help/course-materials/2019/BSS2019/05_Annotations.html -
        org.hs.eg.db

-   Load the data into R.

```{r}
# Create the data folder if it doesn't exist
if (!dir.exists("data")) {
  dir.create("data")
}

# Define the file path to the plots directory
plots_dir <- "plots"

# Create the plots folder if it doesn't exist
if (!dir.exists(plots_dir)) {
  dir.create(plots_dir)
}

# Define the file path to the results directory
results_dir <- "results"

# Create the results folder if it doesn't exist
if (!dir.exists(results_dir)) {
  dir.create(results_dir)
}
```

### [üö®Pause here and load the data.]{.underline}

[Please insert your data files in the data directory such that the
single folder from refine.bio containing the metadata TSV and main TSV
file are present in the immediate subdirectory. This can be downloaded
from google drive,
[here!](https://drive.google.com/file/d/1aOnupOgIn-b7rSoGdRflNBvpfQPcR5v5/view?usp=sharing "Download folder to be unzipped as subdirectory of Data!")]{.underline}

```{r}
# Define the file path to the data directory
data_dir <- file.path("data", "SRP075377")

# Declare the file path to the gene expression matrix file inside directory saved as `data_dir`
data_file <- file.path(data_dir, "SRP075377.tsv")

# Declare the file path to the metadata file inside the directory saved as `data_dir`
metadata_file <- file.path(data_dir, "metadata_SRP075377.tsv")

# Check that files exist and are usable.
file.exists(data_file)
file.exists(metadata_file)
```

[This next step loads data and annotation libraries per the
instructions, but does not attempt to transform our Gene column to the
Entrez IDs as this was unnecessary for downstream
processing.]{.underline}

```{r include=FALSE}
#Install and attach the Homo Sapiens annotation package.
if (!("org.Hs.eg.db" %in% installed.packages())) {
  BiocManager::install("org.Hs.eg.db", update = FALSE)
}
library(org.Hs.eg.db)

# Attach library for pipe (%>%)
library(magrittr)

# Read in data and metadata TSV file and make Gene column into row names
metadata <- readr::read_tsv(metadata_file)
expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")
```

```{r}
# Make the data in the order of the metadata
expression_df <- expression_df %>%
  dplyr::select(metadata$refinebio_accession_code)

# Check if this is in the same order
all.equal(colnames(expression_df), metadata$refinebio_accession_code)

# Bring back the "Gene" column in preparation for mapping
expression_df <- expression_df %>%
  tibble::rownames_to_column("Gene")

# Map Ensembl IDs to their associated Entrez IDs
mapped_list <- mapIds(
  org.Hs.eg.db, # Replace with annotation package for your organism
  keys = expression_df$Gene,
  keytype = "ENSEMBL", # Replace with the type of gene identifiers in your data
  column = "ENTREZID", # The type of gene identifiers you would like to map to
  multiVals = "list"
)

head(mapped_list) #Amazing.
```

```{r include=FALSE}
# Let's make our list a bit more manageable by turning it into a data frame
mapped_df <- mapped_list %>%
  tibble::enframe(name = "Ensembl", value = "Entrez") %>%
  # enframe() makes a `list` column; we will simplify it with unnest()
  # This will result in one row of our data frame per list item
  tidyr::unnest(cols = Entrez)
```

What size is your expression matrix?

```{r}
dim(expression_df)
```

How many genes does it include?

[43363 ensembl IDs OR 44380 Entrez IDs. 13996 Ensembl's had no ENTREZ
IDs, but some mapped to many...]{.underline}

```{r}
summary(as.factor(mapped_df$Entrez), maxsum = 5)
```

```{r}
multi_mapped <- mapped_df %>%
  # Let's count the number of times each Ensembl ID appears in `Ensembl` column
  dplyr::count(Ensembl, name = "entrez_id_count") %>%
  # Arrange by the genes with the highest number of Entrez IDs mapped
  dplyr::arrange(desc(entrez_id_count))

# Let's look at the first 6 rows of our `multi_mapped` object
head(multi_mapped)
```

[Looks like some ensemblIDs mapped to as many as 210 Entrez
IDs...]{.underline}

How much variation do you see in the data? To answer these questions,
log-scale the data, calculate per-gene median expression ranges, then
make a density plot showing those results.

```{r include=FALSE}
#Log scale the data
logscaleData <- expression_df
logscaleData[, -1] <- log(logscaleData[, -1])
logscaleData$median = apply(logscaleData[, -1], 1, median)

#Create a function to find ranges per row
calculate_range <- function(row) {
  max_value <- max(row)
  min_value <- min(row)
  range_value <- max_value - min_value
  return(range_value)
}

#Apply this to find ranges
logscaleData$row_ranges <- apply(logscaleData[,-1], 1, calculate_range)
```

```{r}
justPlotThis <- dplyr::select(logscaleData, columns="Gene", "median", "row_ranges")
justPlotThis <- justPlotThis %>% 
  dplyr::rename(
    Genes = columns,
    Median = median,
    Range = row_ranges,
    )

plot(density(justPlotThis$Median))
```

[The above density plot simply shows the density of the median values.
It is barely bimodal with a second group visible at just over 1, these
genes being the small group that is being expressed more than the
majority.]{.underline}

```{r}
# First let's install the requisite libraries.
    if (!("affy" %in% installed.packages())) {
      # Install this package if it isn't installed yet
      BiocManager::install("affy", update = FALSE)
    }
    
expression_df <- readr::read_tsv(data_file) %>%
  # Tuck away the Gene ID column as row names
  tibble::column_to_rownames("Gene")

library(scales)
library(affy)
myColors <- hue_pal()(4)

plotDensity(log(expression_df+1), col=rep(myColors, each=3),
            lty=c(1:ncol(logscaleData)), xlab='Log(count)',
            main='Expression Density Distribution')
```

#### Summarize your findings.

[The majority of the genes show negative median log-scaled expression (0
to 1), simply implying minimal expression relative to other genes.
However, a group can be seen with values near 1-2. These imply that the
genes in this group are those which, relative to other genes gathered
from the pancreatic islet cells samples, are being expressed the most.
This stands against no particular control, however, and so does not
yield any meaningful conclusions of the genes at hand being diabetic,
but rather just genes that are expressed more.]{.underline}

```{r}
plot(density(justPlotThis$Range))
```

[The range density plot shows an intriguing bimodal distribution. A
large group of genes has a minimal range, whereas the majority of the
genes fall close to a range of 3.5 (log scale) in expression between
samples. Genes with a low variability across all samples are not
interesting since this data set contains both diabetic and control
patients.]{.underline}

```{r include=FALSE}
#Cleanup
rm(justPlotThis)
rm(logscaleData)
rm(mapped_df)
rm(mapped_list)
rm(multi_mapped)
rm(myColors)
rm(calculate_range)
```

### 2. Now that you have loaded the expression data into R, generate a PCA plot:

-   If you have counts file(s), follow these DESeq2 directions to
    generate an expression matrix.
-   Use the DESeq2 function plotPCA() to generate your plot (see here)
-   Color your plot by the 2 groups you identified in assignment 1
    (e.g., Diabetic vs Non-Diabetic)
-   Make sure you include a legend and label the axes!

```{r include=FALSE}
# First let's install the requisite libraries.
if (!("DESeq2" %in% installed.packages())) {
  # Install this package if it isn't installed yet
  BiocManager::install("DESeq2", update = FALSE)
}

# Attach the DESeq2 library
library(DESeq2)
```

Alright, let's prep the metadata to better select what we need.

```{r include=FALSE}
new_expression_df <- readr::read_tsv(data_file) %>%
  tibble::column_to_rownames("Gene")

metadata <- metadata %>%
  dplyr::mutate(diabetes = dplyr::case_when(
    stringr::str_detect(refinebio_subject, "non t2d") ~ "reference",
    stringr::str_detect(refinebio_subject, "t2d") ~ "diabetic",
  ))

#Remove ambiguously-labeled samples from metadata
culledMeta <- metadata[!(metadata$refinebio_subject=="pancreatic islets"),]

discardColumns <- metadata[(metadata$refinebio_subject=="pancreatic islets"),]
discardColumns = as.vector(discardColumns$refinebio_accession_code)
length(discardColumns)
#Preserve only columns in expression_df that match one of the accession ids
culled_expression_df = new_expression_df[,!(names(new_expression_df) %in% discardColumns)]

# Make mutation_status a factor and set the levels appropriately
culledMeta <- culledMeta %>%
  dplyr::mutate(
    # Here we define the values our factor variable can have and their order.
    diabetes = factor(diabetes, levels = c("reference", "diabetic"))
  )
```

```{r}
round_culled_expression_df <- round(culled_expression_df)


ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = round_culled_expression_df,
  # Supply the `colData` with our metadata data frame
  colData = culledMeta,
  # Supply our diabetes variable to `design`
  design = ~diabetes #Diabetes_status
)

dds_norm <- vst(ddset)

plotPCA(
  dds_norm,
  intgroup = "diabetes"
)

pca_results <-
  plotPCA(
    dds_norm,
    intgroup = c("diabetes"),
    returnData = TRUE # This argument tells R to return the PCA values
  )

ddset <- DESeqDataSetFromMatrix(
  # Here we supply non-normalized count data
  countData = round_culled_expression_df,
  # Supply the `colData` with our metadata data frame
  colData = culledMeta,
  # Supply our diabetes variable to `design`
  design = ~diabetes #Diabetes_status
)

dds_norm <- vst(ddset)

plotPCA(
  dds_norm,
  intgroup = "diabetes"
)

pca_results <-
  plotPCA(
    dds_norm,
    intgroup = c("diabetes"),
    returnData = TRUE # This argument tells R to return the PCA values
  )

```

-   PCA Plot Summary

-   PCA Plot Summary The PCA plot shows that the two groups diabetic and
    non-diabetic separate into two different clusters. The non-diabetic
    clusters tends to have less variance and is one uniform cluster
    while the diabetic samples are separated into two different uniform
    clusters that have more variance.

-   If you have 3 or 4 students in your group, also generate either
    t-SNE or UMAP plot, and summarize the differences and similarities
    between your two plots.

-   UMAP:

```{r}
library("umap")

gene <- DESeqDataSetFromMatrix(
  countData = round_culled_expression_df, # Counts values for all samples in our rounded dataset
  colData = culledMeta, #  Supply the `colData` with our metadata data frame
  design = ~diabetes #Supply our diabetes variable to `design`
  # Diabetes_status
)
gene_norm <- vst(gene)
normalized_counts <- assay(gene_norm) %>%
  t() # transpose this data so each row is a sample
umap_results <- umap::umap(normalized_counts)
umap_plot_df <- data.frame(umap_results$layout) %>%
  # Turn sample IDs from row names into a column
  tibble::rownames_to_column("refinebio_accession_code") %>%
  # Add the metadata into this data frame. Match by sample IDs
  dplyr::inner_join(metadata, by = "refinebio_accession_code")
#umap_plot_df (no need to show table, takes up 100 pages on pdf)
ggplot(
  umap_plot_df,
  aes(
    x = X1,
    y = X2,
    color = diabetes
  )
) +
  geom_point() # Plot individual points to make a scatterplot
```

-   Save your plot(s) and summarize your findings.

```{r}
#Cleanup
rm(dds_norm)
rm(gene)
rm(gene_norm)
rm(new_expression_df)
rm(normalized_counts)
rm(pca_results)
rm(umap_plot_df)
rm(umap_results)
rm(ddset)
```

-   UMAP Plot Summary The UMAP plot shows that the two groups are
    clustered together in distinct groups. These clusters are the two
    different cell types, diabetic and reference. There are a few
    outliers in the plot where a few samples are displayed at the bottom
    of the plot.

PCA Plot vs UMAP Plot Comparison - PCA is a linear technique that
captures global structure and is computationally efficient, while UMAP
is a nonlinear technique that excels at preserving local structure,
making it well-suited for exploring complex relationships and
identifying clusters in high-dimensional data.

### 3. Perform differential analysis on the samples from your two groups, following the directions below

-   alexslemonade.github.io/refinebio-examples/03-rnaseq/differential-expression_rnaseq_01.html

-   Create a volcano plot of your data, following the directions above

    ```{r include=FALSE}
    if (!("EnhancedVolcano" %in% installed.packages())) {
      # Install this package if it isn't installed yet
      BiocManager::install("EnhancedVolcano", update = FALSE)
    }
    if (!("apeglm" %in% installed.packages())) {
      # Install this package if it isn't installed yet
      BiocManager::install("apeglm", update = FALSE)
    }

    # Attach the ggplot2 library for plotting
    library(ggplot2)

    # We will need this so we can use the pipe: %>%
    library(magrittr)
    ```

    Volcano plot time.

    ```{r include=FALSE}
    filtered_expression_df <- culled_expression_df %>%
      dplyr::filter(rowSums(.) >= 4000)

    #This culls roughly 35000 genes that show insignificant expression.

    gene_matrix <- round(filtered_expression_df)
    ddset <- DESeqDataSetFromMatrix(
      # Here we supply non-normalized count data
      countData = gene_matrix,
      # Supply the `colData` with our metadata data frame
      colData = culledMeta,
      # Supply our experimental variable to `design`
      design = ~diabetes
    )
    ```

    ```{r}
    #This WILL take some time. To reduce it down from a minute, use adjust the number on the third line of the above cell. The higher, the more genes filtered out, the shorter DESeq will take but the worse the volcano plot will be.
    deseq_object <- DESeq(ddset)
    ```

    ```{r include=FALSE}
    deseq_results <- results(deseq_object)
    deseq_results <- lfcShrink(
      deseq_object, # The original DESeq2 object after running DESeq()
      coef = 2, # The log fold change coefficient used in DESeq(); the default is 2.
      res = deseq_results # The original DESeq2 results table
    )

    # this is of class DESeqResults -- we want a data frame
    deseq_df <- deseq_results %>%
      # make into data.frame
      as.data.frame() %>%
      # the gene names are row names -- let's make them a column for easy display
      tibble::rownames_to_column("Gene") %>%
      # add a column for significance threshold results
      dplyr::mutate(threshold = padj < 0.05) %>%
      # sort by statistic
      dplyr::arrange(dplyr::desc(log2FoldChange))

    volcano_plot <- EnhancedVolcano::EnhancedVolcano(
      deseq_df,
      lab = deseq_df$Gene,
      x = "log2FoldChange",
      y = "padj",
      pCutoff = 0.01
    )

    ```

    ```{r}
    # Print out plot here
    volcano_plot
    ```

-   Create a table of differentially expressed genes.

    ```{r}
    head(deseq_df)
    ```

-   Save and summarize your findings.

    ```{r include=FALSE}
    readr::write_tsv(
      deseq_df,
      file.path(
        results_dir,
        "diff_expr_results.tsv" # Replace with a relevant output file name
      )
    )
    ```

[It seems a handful of genes have stood out as having significant fold
change in bot the positive and negative direction relative to the
control. Of the 6827 genes examined, 357 meet the threshold discovered
for significant change in differential expression, visualized as north
of the grey line towards the bottom of the volcano plot. Some are
exceptionally negative / positive, and are labeled for their
uniqueness.]{.underline}

```{r}
#Cleanup

```

### 4. Extract the list of significantly differentially expressed genes, and generate a heatmap using ComplexHeatmap

-   Package reference
    (<https://jokergoo.github.io/ComplexHeatmap-reference/book/>)

-   Add a side bar colored by sample groupings (cancer vs not, etc.)

```{r include=FALSE}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")

BiocManager::install("ComplexHeatmap")
```

```{r}
#transpose expression_df so it can merge
#extract only the Gene column from data
transposed <- expression_df
transposed <- t(transposed)
transposed <- as.data.frame(transposed)
transposed <- transposed %>%
  tibble::rownames_to_column("Code")
transposed <- transposed[-1,]
```

```{r}
#rename ascession_code in copy of metadata to "Code" for merge to work, and cull unneeded columns
metadata_test <- metadata
names(metadata_test)[names(metadata_test) == "refinebio_accession_code"] <- "Code"
merge_filter = metadata_test[,c("Code","refinebio_subject")]

#merge culled metadata with transposed
merged <- merge(transposed,merge_filter,by="Code")
```

```{r include=FALSE}
#extract ONLY significantly expressed genes
deseq_df_enr <- readr::read_tsv(file.path(results_dir, "diff_expr_results.tsv"))
deseq_df_enr<-deseq_df_enr[(deseq_df_enr$threshold==TRUE),]

merged <- merged[,c("refinebio_subject", deseq_df_enr$Gene)]
```

```{r}
#group by refinebio subject
new_merged <- aggregate(merged, list(merged$refinebio_subject), FUN=mean)
new_merged <- new_merged[order(new_merged$Group.1),]
new_merged <- new_merged[-c(2)] %>% 
  column_to_rownames(var="Group.1")

row_names_df_to_remove<-c("pancreatic islets")
new_merged <- new_merged[!(row.names(new_merged) %in% row_names_df_to_remove),]
```

```{r}

#heatmap

ht = ComplexHeatmap::Heatmap(new_merged,heatmap_legend_param = list(
        title = "Expression", at = c(0, 50, 100,150)
    ),column_title = "Significantly Differentially Expressed Genes Heatmap",show_column_names = FALSE,cluster_columns = FALSE,cluster_rows = FALSE)
ComplexHeatmap::draw(ht)
```

### HEATMAP SUMMARY:

To create this Heatmap, we used ComplexHeatmap and simply aligned our raw data with the significantly expressed gene table, then fed in only the significantly expressed gene data to ComplexHeatmap. We also grouped this data by the subject, with "non t2d" denoting a subject without Type 2 Diabetes and "t2d" denoting a subject with diabetes. We found that ComplexHeatmap's default settings clustered the rows and columns so that they appeared in a different order than how they appear in the input table. To fix this, we set cluster to FALSE for both rows and columns.

Based on this heatmap from our significantly expressed gene data, there does appear to be some degree of difference between how certain groups of genes are expressed across the subjects with diabetes and those without. Interestingly, there appear to be differences in how diabetes is expressed within the diabetes groups, with subjects 1,2, and 3 differing from subjects 4,5, and 6. A similar division within the non-diabetes group, if it exists, does not appear to be quite as apparent.


### 5. Run enrichment analysis on your data using your selected method and ontology

Extract the list of deferentially expressed genes and run enrichment
analysis.

```{r include=FALSE}
#Extracting differentially expressed genes:
deseq_df_enr <- readr::read_tsv(file.path(results_dir, "diff_expr_results.tsv"))
deseq_df_enr<-deseq_df_enr[(deseq_df_enr$threshold==TRUE),]

readr::write_tsv(
  deseq_df_enr,
  file.path(
    results_dir,
    "diff_expr_results_enr.tsv" # Replace with a relevant output file name
  )
)


```

Each student in your team should run a different method OR ontology
(e.g., if there are 4 students on the team, there should be results for
4 applications in your assignment writeup).

-   Choose a method:

    -   gProfiler2

[We have opted to use gprofiler2 to avoid the challenges associated with
writing R and manipulating our already conclusive data. All three of us
are exploring three different ontologies: Biological pathways (<GO:BP>),
Molecular Function (<GO:MF>), and Cellular Components (<GO:CC>). The
above code isolates only statistically significant genes which we can
then paste into gprofiler/gost. Seeing as part 5 contains no further
instructions, see part 6 for corresponding write up.]{.underline}

### 6. Create a table of the enriched processes found for each method in step 4 (one table per method). Create a table showing statistically significantly enriched terms and any characteristics shared by the method you used (e.g., q-value, p-value, log fold change)

![im3](https://cdn.discordapp.com/attachments/1148297410318176256/1156772921298124830/gProfiler_hsapiens_2023-09-28_01-39-56.png)

[The above table was produced using gost, an online tool that attempts
to display genes clustered by various ontologies, ontologies being such
significance as the molecule to which they contribute, regulatory
effects they perform, or relationship with a disease among other
categories. Of signficance is cytoplasmic relationship within
particularly the <GO:CC> table. We can see a very significant P value
here of 1.324x10\^-14, which is seemingly reasonable since diabetes
would influence certain metabolic pathways naturally present in the
cytoplasm, or so we would assume from the results. Biological quality
regulation was also closely tied to the differential expression of our
genes, as well as protein binding, this last one falling under molecular
function. This is intriguing as it may relate to insulin receptors'
capacity to bind free glucose, though it is our understanding that this
is a fundamental quality not only of pancreatic islet cells but of all
somatic cells participating in metabolism.]{.underline}

### 7. Write a short summary to go with each plot/table you create. Describe what you did, what parameters you used (if any) and an interesting result from it.

## As a team, fill out the team evaluation table below.

![im1](https://cdn.discordapp.com/attachments/1148297410318176256/1156784591059816558/image.png)
![im1](https://cdn.discordapp.com/attachments/1148297410318176256/1156784591059816558/image.png)

![im2](https://cdn.discordapp.com/attachments/1148297410318176256/1156784591059816558/image.png)
![im2](https://cdn.discordapp.com/attachments/1148297410318176256/1156784591059816558/image.png)

### 8. Combine all into a single file and submit on Canvas. Make sure that all your code is added to your GitHub repository.

# ‚úÖ
